### PR TITLE
bugfix: parallel request handle throw IndexOutOfBoundsException

### LINKED ISSUE #5280: java.lang.IndexOutOfBoundsException: Index: 1, Size: 0
- [ ] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
压测中出现,批量rm注册分支后,并行处理时数组越界
```
2023-02-01 20:33:18.344 ERROR --- [   ServerHandlerThread_1_6_500] i.s.c.r.p.s.ServerOnRequestProcessor     : handle request error: java.lang.IndexOutOfBoundsException: Index: 1, Size: 0
==>
java.util.concurrent.ExecutionException: java.lang.IndexOutOfBoundsException: Index: 1, Size: 0
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at io.seata.core.rpc.processor.server.ServerOnRequestProcessor.onRequestMessage(ServerOnRequestProcessor.java:201)
	at io.seata.core.rpc.processor.server.ServerOnRequestProcessor.process(ServerOnRequestProcessor.java:123)
	at io.seata.core.rpc.netty.AbstractNettyRemoting.lambda$processMessage$2(AbstractNettyRemoting.java:280)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.IndexOutOfBoundsException: Index: 1, Size: 0
	at java.base/java.util.concurrent.CopyOnWriteArrayList.add(CopyOnWriteArrayList.java:451)
	at io.seata.core.rpc.processor.server.ServerOnRequestProcessor.lambda$onRequestMessage$1(ServerOnRequestProcessor.java:191)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
<==

```

### Ⅱ. Describe what happened

  If there is an exception, please attach the exception trace:

```
Just paste your stack trace here!
```

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. xxx
2. xxx
3. xxx

### Ⅵ. Environment:

- JDK version :
- Seata version:  
- OS :
- Others:

### PR DESCRIPTION
**Pull request by a364176773**
- [ ] I have registered the PR [changes](https://github.com/seata/seata/tree/develop/changes).

### Ⅱ. Does this pull request fix one issue?

fixes #5280

### CODE REVIEW

File: core/src/main/java/io/seata/core/rpc/processor/server/ServerOnRequestProcessor.java [hunk 1]
```diff
@@ -179,26 +178,26 @@ private void onRequestMessage(ChannelHandlerContext ctx, RpcMessage rpcMessage)
                     }
                 }
             } else {
-                List<AbstractResultMessage> results = new CopyOnWriteArrayList<>();
-                List<CompletableFuture<Void>> completableFutures = null;
+                List<AbstractResultMessage> results = new ArrayList<>();
```
Comment (a364176773) [inline core/src/main/java/io/seata/core/rpc/processor/server/ServerOnRequestProcessor.java:181]: 不在多线程中add，不存在线程安全问题，改为ArrayList

File: core/src/main/java/io/seata/core/rpc/processor/server/ServerOnRequestProcessor.java [hunk 2]
```diff
@@ -179,26 +178,26 @@ private void onRequestMessage(ChannelHandlerContext ctx, RpcMessage rpcMessage)
...
+                List<CompletableFuture<AbstractResultMessage>> completableFutures = null;
```
Comment (a364176773) [inline core/src/main/java/io/seata/core/rpc/processor/server/ServerOnRequestProcessor.java:182]: 每个future的返回值都为AbstractResultMessage

File: core/src/main/java/io/seata/core/rpc/processor/server/ServerOnRequestProcessor.java [hunk 3]
```diff
@@ -179,26 +178,26 @@ private void onRequestMessage(ChannelHandlerContext ctx, RpcMessage rpcMessage)
...
                     if (PARALLEL_REQUEST_HANDLE) {
                         if (completableFutures == null) {
                             completableFutures = new ArrayList<>();
                         }
                         int finalI = i;
-                        completableFutures.add(CompletableFuture.runAsync(() -> {
-                            results.add(finalI, handleRequestsByMergedWarpMessage(
-                                ((MergedWarpMessage)message).msgs.get(finalI), rpcContext));
-                        }));
+                        completableFutures.add(CompletableFuture.supplyAsync(() -> handleRequestsByMergedWarpMessage(
```
Comment (a364176773) [inline core/src/main/java/io/seata/core/rpc/processor/server/ServerOnRequestProcessor.java:189]: 有序的将每个执行request处理的future放入completableFutures

File: core/src/main/java/io/seata/core/rpc/processor/server/ServerOnRequestProcessor.java [hunk 4]
```diff
@@ -179,26 +178,26 @@ private void onRequestMessage(ChannelHandlerContext ctx, RpcMessage rpcMessage)
...
                     } else {
                         results.add(i,
                             handleRequestsByMergedWarpMessage(((MergedWarpMessage)message).msgs.get(i), rpcContext));
                     }
                 }
                 if (CollectionUtils.isNotEmpty(completableFutures)) {
                     try {
-                        CompletableFuture.allOf(completableFutures.toArray(new CompletableFuture[0])).get();
+                        for (CompletableFuture<AbstractResultMessage> completableFuture : completableFutures) {
+                            results.add(completableFuture.get());
```
Comment (a364176773) [inline core/src/main/java/io/seata/core/rpc/processor/server/ServerOnRequestProcessor.java:199]: 按顺序读取再加回去，completableFutures的数量等于msgs，顺序也一致

**Review by wangliang181230 — APPROVED** [review]
LGTM

**Review by xingfudeshi — APPROVED** [review]
LGTM

**Pull request merged**